homeassistant/components/yolink/__init__.py

```diff
@@ -0,0 +1,122 @@
+
+from . import api, config_flow
+from .const import DOMAIN, HOME_ID, HOME_SUBSCRIPTION
+
+SCAN_INTERVAL = timedelta(minutes=5)
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_CLIENT_ID): cv.string,
+                vol.Required(CONF_CLIENT_SECRET): cv.string,
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+PLATFORMS = ["sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the yolink component."""
+    hass.data[DOMAIN] = {}
+    if DOMAIN not in config:
+        return True
+
+    config_flow.OAuth2FlowHandler.async_register_implementation(
+        hass,
+        config_entry_oauth2_flow.LocalOAuth2Implementation(
+            hass,
+            DOMAIN,
+            config[DOMAIN][CONF_CLIENT_ID],
+            config[DOMAIN][CONF_CLIENT_SECRET],
+            OAUTH2_AUTHORIZE,
+            OAUTH2_TOKEN,
+        ),
+    )
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up yolink from a config entry."""
+    implementation = (
+        await config_entry_oauth2_flow.async_get_config_entry_implementation(
+            hass, entry
+        )
+    )
+
+    session = config_entry_oauth2_flow.OAuth2Session(hass, entry, implementation)
+    auth_mgr = api.ConfigEntryAuth(
+        hass, aiohttp_client.async_get_clientsession(hass), session
+    )
+    yolink_http_client = YoLinkClient(auth_mgr)
+    yolink_mqtt_client = MqttClient(auth_mgr)
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        "client": yolink_http_client,
+        "mqttClient": yolink_mqtt_client,
+        "devices": [
+            {
+                "deviceId": "",
+                "name": "",
+                "type": "",
+                "token": "",
+            }
+        ],
+    }
+    try:
+        async with async_timeout.timeout(5):
+            home_response = await yolink_http_client.get_general_info()
+            if not (home_response.data["id"] is None):
+                hass.data[DOMAIN][entry.entry_id][HOME_ID] = home_response.data["id"]
+                hass.data[DOMAIN][entry.entry_id][HOME_SUBSCRIPTION] = HomeEventMqttSub(
+                    home_response.data["id"]
+                )
+                yolink_mqtt_client.subHome(
+                    hass.data[DOMAIN][entry.entry_id][HOME_SUBSCRIPTION]
+                )
+        async with async_timeout.timeout(5):
+            devices_response = await yolink_http_client.get_auth_devices()
+            if not (devices_response.data["devices"] is None):
+                hass.data[DOMAIN][entry.entry_id]["devices"] = devices_response.data[
+                    "devices"
+                ]
+        await yolink_mqtt_client.async_connect()
+    except BaseException as err:
+        _LOGGER.warning("Call yolink api failed: %s", err)
```

Comment:
Are there more errors possible?
- What if authentication is incorrect, how will that be raised?
- Can the client recover from a base exception by retrying? Than you could raise `ConfigEntryNotReady`.

Reply:
client will recover from retrying,  If the authentication fails, the process will not go down